### PR TITLE
west/sign: Rework of how MCUboot partitions are taken from DT

### LINKED ISSUE #54373: Mcuboot swap type is `test` when update fails
**Describe the bug**
When installing an image with mcuboot swap using move, the installation can fail if the image is too big.
In that case the old image is loaded again.
If we ask the image for its swap type using the `mcuboot_swap_type`-function we expect it to say `none`, but we get `test` instead.

```
#include <stdbool.h>
#include <dfu/dfu_target.h>
#include <dfu/dfu_target_mcuboot.h>
#include <dfu/mcuboot.h>

void main(void)
{
    mcuboot_swap_type();
}
```

**To Reproduce**
1. Make mcuboot update swap using move fail by providing an image that is too big
2. use the `mcuboot_swap_type();` function somewhere in the firmware.

**Expected behavior**
After the upgrade failed, mcuboot reverts to the previous image. It is expected to be of swap type `none` but instead it is of type `test`

**Impact**
If someone is using the flag to check for a "healthy" boot, they get unexpected behavior. Reset cycles for example.

**Logs and console output**
```
*** Booting Zephyr OS build v3.1.99-ncs1-12-gaa84ba70d601  ***
<inf> mcuboot_util: Swap type: none
```

Let me know if additional context is needed.

### PR DESCRIPTION
**Pull request by de-nordic**
The commit moves from using chosen `zephyr,flash` to flash device where chosen `zephyr,code-partition` is placed. There has been special class created MCUbootDTConfig that is helper class for accessing MCUboot type partitions. The class incorporates edt_flash_node logic, and the method has been removed.

Object of the MCUbootDTConfig class is now used by mcuboot_swap_validate method, which replaces edt_flash_params, that is supposed to validate partition layout for swap configuration. mcuboot_swap_validate extends edt_flash_params with additional checks and warning logs.
The mcuboot_swap_validate maintains compatibility with legacy behaviour where single slot MCUboot configuration is assumed if slot1_partition is not found.

**Note**: This is base work for separating current validation into swap-move and single application mode, and addition of DirectXIP and swap-scratch.

- [x] Documentation is missing
- [x] Commit with binary file validation against image slot size
- [x]  Separate swap-move and single application mode
- [x]  Support for swap-scratch
- [x]  Support for DirectXIP
- [x] Release notes 

**CONTENTS**
The PR consists of commits that:
 1) rewrite swap/single slot validation
 2) add application size validation against slot size
 3) split swap/single validation to separate swap and separate single slot validation
 4) add esp32 overlay required 
 5) Kconfig options to let application know what mode of operation has been set for MCUboot
 6) swap-scratch validation support

The sing script commits can be squashed, they are separated to make it easier to understand steps taken.

Fixes #54373.

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
@de-nordic the pull request description is doing a great job of saying what is happening, but not why. Why is this change needed?

File: scripts/west_commands/sign.py [hunk 1]
```diff
@@ -205,6 +205,142 @@ def sign(self, command, build_dir, build_conf, formats):
...
+        other = mcbdtconf.get_partition('slot1_partition')
+
+        if other:
+            op_addr, op_size = mcbdtconf.get_partition_params(other)
+            # Always go with smaller partition size, as swap will not be possible
+            if op_size != cp_size:
+                log.wrn(f"slot0_partition and slot1_partition differ in size:"
+                             " ${cp_size} vs ${op_size}. Smaller size will be used.")
+            if op_size < cp_size:
+                cp_size = op_size
```
Comment (nordicjm) [inline scripts/west_commands/sign.py:338]: Indent one extra

**Comment by de-nordic** [comment]
> @de-nordic the pull request description is doing a great job of saying what is happening, but not why. Why is this change needed?

The MCUbootDTConfig  has been provided to make it easier to get MCUboot related partition configuration, to use with series of `mcuboot_swap_validate`  type methods: `mcuboot_single_validate`, `mcuboot_scratch_validate` and `mcuboot_directxip_validate` . These will validate build app vs DT partitions, assuming MCUboot configuration, and will either refuse singing or return proper parameters for signing, eventually giving some warnings (for example single configuration will warn that there are separate slot0 and slot1, which probably wastes space).
Above functions are missing now, but to be able to reuse parts of edt_ methods code, I have decoded to move them to MCUbootDTConfig  class.

**Comment by mbolivar-nordic** [comment]
@de-nordic sorry I was unclear -- why is the change of chosen node needed? Won't this break existing users?

**Comment by de-nordic** [comment]
> @de-nordic sorry I was unclear -- why is the change of chosen node needed? Won't this break existing users?

No it should not.
While signing image we expect it to be selected by `zephyr,code-partition`, so we should also figure out flash the `zephyr,code-partition` is placed on, instead of using `zephyr,flash`.
If user has previously set `zephyr,code-partition` to partition placed on something different than `zephyr,flash` than it would probably not work for the user anyway. So user would end up correcting `zephyr,flash` to point to flash where `zephyr,code-partition` is placed, or moving `zephyr,code-partition` to that flash.
Basically in previous code we were a little bit out-of-sync with DTS: we would expect `zephyr,code-partition` to point to binary we start with, but use `zephyr,flash` to sign image (which have to be the same for this to work anyway). Common stuff we check was slot labels.

**Review by mbolivar-nordic — DISMISSED** [review]
The explanation makes sense to me. Thank you.

Reassigning the PR since @nordicjm has started taking overall responsibility for flashing, but it LGTM.

**Comment by de-nordic** [comment]
> The explanation makes sense to me. Thank you.
> 
> Reassigning the PR since @nordicjm has started taking overall responsibility for flashing, but it LGTM.

Thanks.

Anyway, it seems to break esp32 which I need to take look into. It seems that they select CONFIG_USE_DT_CODE_PARTITION, vaie CONFIG_BOOTLOADER_MCUBOOT, even though they do not use th zephyr,code-partition at all. This is misleading.
I also found that they use sometimes `zephyr,flash` but also seem to hardcode, in Kconfigs, specific flash device via DT path.

File: samples/boards/esp32/flash_encryption/boards/esp32.overlay [hunk 2]
```diff
@@ -7,3 +7,9 @@
 &flash0 {
```
Comment (nordicjm) [inline samples/boards/esp32/flash_encryption/boards/esp32.overlay:7]: typo in this commit: `samples/boards/esp32: Add chosen zpehyr,code-partition to overlay` *zephyr

File: scripts/west_commands/sign.py [hunk 3]
```diff
@@ -237,6 +237,264 @@ def sign(self, command, build_dir, build_conf, formats):
...
+        cp_addr, cp_size = code.get_parameters()
+
+        other = mcbdtconf.get_partition('slot1_partition')
+
+        if other:
+            _, op_size = other.get_parameters()
+            # Always go with smaller partition size, as swap will not be possible
+            if op_size != cp_size:
+                log.wrn("slot0_partition and slot1_partition differ in size:"
+                        f" ${cp_size} vs ${op_size}. Smaller size will be used.")
```
Comment (nordicjm) [inline scripts/west_commands/sign.py:582]: Have seen some people use different slot sizes because of the trailer information stored on one slot only, might be an idea (in future) to check which slot is larger and ignore the error if it's that one

File: scripts/west_commands/sign.py [hunk 4]
```diff
@@ -237,6 +237,264 @@ def sign(self, command, build_dir, build_conf, formats):
...
+                if op_size < cp_size:
+                    cp_size = op_size
+        else:
+            log.wrn('slot1_partition not found, assuming single slot configuration')
```
Comment (nordicjm) [inline scripts/west_commands/sign.py]: I think this might become a bit verbose to see this every time if the user is using single slot mode, debug level log might be better suited

File: scripts/west_commands/sign.py [hunk 5]
```diff
@@ -12,6 +12,7 @@
 import shlex
```
Comment (nordicjm) [inline scripts/west_commands/sign.py:12]: typo in commit: size, expanded by size off required additions, is compared agains
*against

File: scripts/west_commands/sign.py [hunk 6]
```diff
@@ -422,6 +423,20 @@ def get_code_partition(self):
     def __init__(self, quiet=False):
         self.quiet = quiet
 
+    @staticmethod
+    def mcuboot_magic_size(flash_alignment=8):
+        '''Get MCUboot magic size
+
+        Magic size is is 16 octets aligned to flash write alignment (min. 8) + 3 flags,
```
Comment (nordicjm) [inline scripts/west_commands/sign.py:430]: -is

**Review by nordicjm — CHANGES_REQUESTED** [review]
(not fully reviewed)

File: scripts/west_commands/sign.py [hunk 7]
```diff
@@ -490,11 +505,44 @@ def mcuboot_swap_validate(mcbdtconf, bin_size, quiet=False):
...
+
+            bin_size_wm = bin_size + __class__.mcuboot_magic_size(write_alignment)
+
+            # Binary size can not be greater than slot size without last page,
+            # as the last page is used for move algorithm.
+            if bin_size_wm > (cp_size - erase_size):
+                log.die("Application image with added MCUboot magic will not fit"
+                        " in application slot, for swap-move configured MCUboot.\n"
+                        f"Binary size is {bin_size}, with magic it is {bin_size_wm}.\n"
+                        f"Max allowed slot size is {cp_size} - {erase_size}")
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply scripts/west_commands/sign.py:517]: Didn't see that a Kconfig has been added (in a later commit), this selection probably needs updating depending upon if a scratch partition is present, and be linked with sysbuild. For example, most stm32 boards should be using swap using scratch by default, they cannot operate in swap using move mode

[the remark this answers is not in the document]
Comment (nordicjm) [reply scripts/west_commands/sign.py:517]: Ping @tejlmand for note on sysbuild integration

File: scripts/west_commands/sign.py [hunk 8]
```diff
@@ -520,6 +520,81 @@ def mcuboot_swap_validate(mcbdtconf, bin_size, quiet=False):
...
+
+        if 'slot0_partition' not in code.labels:
+            log.die("west sign, for MCUboot in swap-scratch mode, expects DT chosen"
+                    " zephyr,code-partition to point to node labeled"
+                    " slot0_partition. Currently selected node is labeled"
+                    " " + str(code.labels))
+
+        if 'slot1_partition' and 'scratch_partition' in code.labels:
+            log.die("chosen zephyr,code-partition may not be slot0_partition,"
+                    " slot1_partition and scratch_partition at the same time")
```
Comment (nordicjm) [inline scripts/west_commands/sign.py:550]: This cannot happen

File: scripts/west_commands/sign.py [hunk 9]
```diff
@@ -652,6 +652,83 @@ def mcuboot_single_validate(mcbdtconf, bin_size, quiet=False):
...
+                    " configured for DirectXIP mode not for single application"
+                    " mode?")
+
+        if not dfu.is_internal():
+            log.die("f{dfu_label} is on external flash.\n"
+                    "MCUboot is not yet able to DirectXIP from external flash")
+
+        if mcbdtconf.get_partition('scratch_partition'):
+            log.wrn("scratch_partition not needed for MCUboot in single applicaiton"
+                    " mode")
```
Comment (nordicjm) [inline scripts/west_commands/sign.py:703]: Wrong comment

File: scripts/west_commands/sign.py (hunk 8 above)
Comment (de-nordic) [reply scripts/west_commands/sign.py:550]: Unless somebody does two labels to the same node. I can remove these checks, if not wanted.

File: scripts/west_commands/sign.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (de-nordic) [reply scripts/west_commands/sign.py:517]: This part is for swap-move only, scratch validation is done by separate function.

> For example, most stm32 boards should be using swap using scratch by default, they cannot operate in swap using move mode

OK, so defconfigs for these boards would need to be altered.

File: scripts/west_commands/sign.py [hunk 10]
```diff
@@ -237,6 +237,264 @@ def sign(self, command, build_dir, build_conf, formats):
...
+            if not node:
+                log.die("devicetree has no chosen zephyr,code-partition node;",
+                        " can not validate application partition")
+
+            partition = ImgtoolSigner.DTPartition(node)
+
+            if not partition.is_internal():
+                log.die("west sign currently does not support signing image"
+                        " built for code-partition that has not been placed"
+                        " on internal flash")
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply scripts/west_commands/sign.py:419]: The board is complete, logic for checking whether partition is internal is broken, as it does not assume that there might be more than one compatible assigned to flash node.
